### ISSUE TITLE
Convert Portable Class Library into .NET Standard project

### ISSUE THREAD

**Issue by amoerie** [fo-dicom/fo-dicom#870]
**Suggestion**
Remove DICOM.Portable in favor of having a .NET Standard DICOM project that can be referenced. 
See https://docs.microsoft.com/en-us/dotnet/standard/net-standard for the compatibility table.
This would probably merit a new major release version.

**Reason**
- When we want to add support for TLS, this requires cryptography namespaces that are not available in PCL, See https://github.com/fo-dicom/fo-dicom/issues/688
- Microsoft has deprecated PCL in favor of .NET Standard libraries. This means for example that the Visual Studio 2019 build tools don't even include support for portable class libraries, and probably never will. See https://developercommunity.visualstudio.com/content/problem/349199/microsoftportablev45microsoftportablecsharptargets.html
- PCL can still be built with a full Visual Studio installation, but this is a considerable downside for environments like Docker, where installing a full Visual Studio is cumbersome at the very least.
- PCL generally has a much more limited API available, restricting fo-dicom in its progress.

**Comment by amoerie** [comment]
See https://montemagno.com/how-to-convert-a-pcl-library-to-net-standard-and-keep-git-history/ for guidance

**Comment by gofal** [comment]
Yes, thats the plan. I already postet this plan upon some questions regarding .net standard.
fo-dicom is the only dicom library built in portable. I have no telemetry data so I dont know how much the portable is actually in use, but the plan is, to get a final stable bug-free version of the portable, mono, android, etc versions. 
Maybe you noticed, that there are no big enhancements the last half year but mainly bugfixes in the issues on github. Thats is the reason. Your reimplementation of dicomclient is a kind of exception, because the old dicomclient never got stable so far :-)
Once done soon, there will be a single .net standard project instead of the shared project (some other forks already did this and it worked fine, so it does not have to be done from scratch, there is already some experience) and only for Imaging and Codes there will be platform specific projects that are loaded.

**Comment by amoerie** [comment]
Great, completely agree!

Are you saying that you - or someone else - are already working on this?

**Comment by gofal** [comment]
I am observing the changes in the forks of fo-dicom to get an idea of how other people use fo-dicom. And I saw several with commits like "change to .net standard" and so on. So I knwo there are some users out there who have done this successfully. So there are some who I can ask for experience or other input. I am sure it won't be an difficult task, since the current Dicom.Core project (built to use in .net core) is built as .net standard 1.3, but I can get rid of fo-dicoms proprietary platform specific code like `IOManager` or `IFileReference` and so on. and many compiler conditionals can be removed. and features and enhancements of net standard can be used.

**Comment by gofal** [comment]
I already have written it somewhere, but I dont know where, so I repeat it here. 
I am convinced that the new project should be netStandard 2.0 and not netStandard 2.1 since it will not be supported in Net Framework any more. Net framework is still a very importand platform to run.

**Comment by amoerie** [comment]
Well I for one am very interested in this, as it would make CI a lot easier and generally make my life quite a bit easier in the fork my company is running. Shall I start a PR on this so that perhaps we can work on this collaboratively, or do you already have a branch somewhere to start from?

**Comment by gofal** [comment]
If a netstandard library makes your CI easier, then you can use the dicom.core project instead the dicom.desktop. The dicom.core is compiled as netstandard and therefore also runs on machines without NetCore installed but any other platform supporting netstandard1.3.
I would like to wait until the stable versions are out on nuget before starting this very big refactoring project. Merging could be hard if we start the netstandard too early and there are still some bugs to fix. and maintaining dupacted code is also something I really want to avoid.

**Comment by amoerie** [comment]
Do you mean that you want to wait until 4.0.2 is released, or what exactly do you mean by stable nuget releases?

**Comment by amoerie** [comment]
Circling back to this issue, I think that a Fellow Oak DICOM release to be completely bug free is a utopia that you can wait for for a long time.

The switch to .NET Standard 2.0 would only break fo-dicom for people stuck on .NET Framework < 4.6.1, which is already 5 years old. And people that are stuck on .NET Framework 4.5 will already see a lot of other libraries that no longer provide updates/releases for their platform version. 

I'm not familiar with other platforms that only work with PCL but not with .NET Standard, I assume this would be limited to silverlight/windows phone/..., all platforms without a future as far as I'm aware. 

In my opinion, moving the PCL class libraries to .NET Standard 2.0 is a relatively safe move that will facilitate a lot of scenarios, and now's a good time as any to do this. 

I'm willing to make a PR for this too.

To summarize, here is a list of pros vs cons:

Pros:
- .NET Standard 2.0 has a much broader range of available APIs than PCL. For example, the cryptography APIs (not available in PCL) would enable us to implement SSL with certificates in the DICOM.Network layer.
- .NET Standard 2.0 is easily available on many CI providers and would make continuous integration for Fellow Oak DICOM a lot easier. For example, Azure Devops does not support PCL at all today, which means Fellow Oak DICOM cannot be built there.
- .NET Standard 2.0 is the road forward for Microsoft and platform support is excellent. PCL is now deprecated.
- We might be able to reduce the amount of platform specific code in DICOM.Desktop etc, because .NET Standard 2.0 provides more APIs.
- We might be able to remove the .shproj project in favor of a simple .NET Standard library, drastically simplifying the solution setup in Fellow Oak DICOM.

Cons:
- Consumers stuck on .NET Frameworks older than 4.6.1 will no longer be able to receive updates from Fellow Oak DICOM.
- Consumers stuck on other, old platforms that do not support .NET Standard 2.0 will also no longer be able to receive updates. These include Silverlight, Windows Phone and perhaps older Xamarin versions.

**Comment by gofal** [comment]
Completely agree. I do not want to wait until fo-dicom is completely bug free, but I wanted to have the blocking known bugs fixed (as there were some rendering-issues, the instability with DicomClient that you solved, etc).
I'm not hesitating because I think that the old platforms as silverlight, winphone etc will have a big comeback in near future :-) 
The idea of switching to netstandard is not to only set the compiler options to netstandard - in that case you are already done by using the fo-dicom.netcore project which is built in netstandard 1.3. The idea is to actually use the new api, to do quite some refactorings, remove code and use new classes etc. This means a big construction zone. Even if this is should not be dangerous, there is always the danger that we are forgetting something, that the new api does not behave exactly the same way and we need to change some logic etc. 
customers are using this library for medical computing out there. And I see reliability and stability as far more important in medical computing than having the latest platforms implemented. 

As said I totally agree with you and I myself are looking forward with joy to switch to netstandard, but I hope you understand my reasons. I will start the transition soon.

**Comment by amoerie** [comment]
I understand your reservations @gofal, and I share many of them. I happen to work for a medical company too!

I'll be looking forward to the transition effort, and if I can help in this matter, please contact me.

**Comment by gofal** [comment]
I started moving the project to netstandard. The code is deployed in branch netstandard.
Things are doing fine so far. mainly I cleaned up code and removed obsolete things, I implemented an DI container to register implementations instead of the XXXManager.SetImplementation(..) syntax.
So rather straight forward things.

A big things are the codecs. I plan to include the codecs that we have as managed classes (RLE, Jpeg baseline, Jpeg2k 8bit, and decompression of JpegLossless) by default. Efferent did such a great work to make the native codecs work on desktop and linux, so I prefer to use their solution and get rid of this in fo-dicom.
And because in the efferent-project there has to be a new interface-implementation done (I will make a pull request there), I wonder if there is a better more generic way to register some codecs in fo-dicom. Currently Efferent had to implement an TranscoderManager and then had to do all the loading of codecs, decide which codec to use etc. 
I am looking for a way where this all still stays within fo-dicom, but you only register an Encoder/Decoder for a special TransferSyntax. 
What is your opinion?

**Comment by amoerie** [comment]
> I started moving the project to netstandard. The code is deployed in branch netstandard.

Excellent! 

> A big things are the codecs. I plan to include the codecs that we have as managed classes (RLE, Jpeg baseline, Jpeg2k 8bit, and decompression of JpegLossless) by default. Efferent did such a great work to make the native codecs work on desktop and linux, so I prefer to use their solution and get rid of this in fo-dicom.

I saw the work done by the Efferent team, and it is very promising work indeed, but I do have some reservations about this:

- fo-dicom does not have many automated tests that verify the correct workings of the codecs. It will be very hard to verify that we don't introduce major breaking changes by switching to another set of codecs. If we make this move, I would very much like to add some tests that take sample DCM files through the codecs and verify the output (by comparing the derived PNGs for example). 

- If you look at [their releases page](https://github.com/Efferent-Health/Dicom-native/releases) they seem to have already set something up with regards to automated testing, but sadly most of the tests seem to fail. If tests fail and nobody fixes the situation, they no longer mean anything!

- Have you communicated with them about this? There is no license information available on their github page, so if we want to incorporate their code into Fellow Oak DICOM they will have to agree to the MS PL license of fo-dicom. 

- Or is the intention to just take a NuGet dependency on their library? In this case, I'm not sure if it's a good idea to leave such a core element of this library in the hands of others. I would much rather have these codecs embedded in the fo-dicom source code, so any fo-dicom contributors can propose changes and so that project is self-contained, in the event that Efferent would suddenly disappear.

> I wonder if there is a better more generic way to register some codecs in fo-dicom. Currently Efferent had to implement an TranscoderManager and then had to do all the loading of codecs, decide which codec to use etc.
> I am looking for a way where this all still stays within fo-dicom, but you only register an Encoder/Decoder for a special TransferSyntax.

I completely agree that we should probably make the "plugin" parts a little more fine-grained, so consumers can make "small cuts" in the logic rather than having to plug in the entire transcoder infrastructure. This should lower the threshold for customising fo-dicom by lowering the risk of doing so.

> What is your opinion?

These are all great ideas and big steps forward! If you would like some support while developing this, I am always willing to review changes or add my own contributions if you would like. Do you have a list of tasks or remaining things to do written down somewhere? We could collaborate on this. :-)

**Comment by amoerie** [comment]
@gofal I see you're making steady progress on this, great! 
Please let me know if I can help in any way.

**Comment by gofal** [comment]
Great, thanks.
I assume I have to merge the code from netstandard-branch into development-branch so that you can do pull requests?
As you can see the netstandard-projects are additional code to the current version, so at the moment  bugs etc have to be fixed twice. Plan is, that I will release 4.0.4 now and soon will release the netstandard version as 5.0.0-alpha. The big thing that is blocking the prerelease is the codec-stuff.

To confess, I am not able to fix anything within the jpeg codecs. So if anyone reported a bug I dont know if anyone in the community could fix it. The guys from efferent seem to do a better job than any other contributor of fo-dicom. So using their stuff is the only possibility I see.

Then netstandard offers the chance to do some changes that are deep in internal and may be risky in normal circumstances. I have two things in mind:

- I do not like the way DicomItem store the values internally. When reading a file, then the binary streams are kept in memory. Then when adding some Items to the DicomDataset, the values are serialized to binary imediatelly. This has the big disadvantage that encodings have to be set as parameter in Add-method. That is dangerous when then there is a dicomtag (0008,0005) this will not be evaluated. I would like to deserialize into managed datatypes on accessing dicomtags and then on sending/writing the dicomdataset into a stream I would serialize those tags.

- The buffer-handling in FellowOakDicom.IO.Buffer should be reworked. The interface IByteBuffer has a method `GetByteRange`. this method creates a new array and copies values. So when reading the stream and parsing, reading etc this method is called very often. There is only read-access, but the data is copied so many times. Use better classes here to reduce memory/GC.

**Comment by amoerie** [comment]
Ok, I agree that working with the Efferent codecs might be the best path forward. I'll reach out to them to start up a discussion, see how they are feeling about this and what the cooperation could look like. We could add their repository as a git submodule, or we could  make a copy of the code here, or we could add them as a NuGet dependency, ... what do you think is best? I think the answer depends on their cooperation.

Concerning big refactorings, I would propose the following roadmap:

1. Get a working version of fo-dicom running on .NET Standard 2.0 ASAP. I would try to keep big refactorings to a minimum for the initial 5.0 release to keep things as stable as possible. We should make a list of things that are required for the 5.0 release, and things that can be postponed until after. 

My required list for 5.0 would look a little something like this:
- Add test suite to verify correct codec input/output using sample images
- Efferent codecs
- Removal of Dicom.Shared.shproj
- Any changes necessary to support the same amount of platforms we support today.

2. Secondly, we could push out a release that includes big refactorings that are now possible thanks to .NET Standard. Depending on whether we break the public API or not, this release could be named 5.1 or 6.0 (does Fellow Oak DICOM adhere to semantic versioning?)

My list for 5.1 or 6.0 would look a little something like this:
- Deserialize values to managed datatypes like you described
- Buffer handling like you described
- Add SSL support to DicomClient / DicomService because .NET Standard includes cryptography libraries, so we can now add it
- Change AppVeyor to something like Github Actions (preferred), Azure Devops, ...
- Add benchmarks for opening and interpreting DicomFiles. This is still quite slow.
-  Revisit DicomService and see where we can make significant improvements now that we have the API surface of .NET Standard 2.0 (Think span & friends)
- ... etc

**Comment by amoerie** [comment]
I've started a discussion in the Efferent codecs repository here: https://github.com/Efferent-Health/Dicom-native/issues/6

I hope you agree with the way that I've phrased our plans, if not feel free to add your own comments.   Let's see if/how they respond.

**Comment by gofal** [comment]
Thanks for your valueable input!
I cannot see the benefit for your 5.0 version. Thats exactly the current Platform-Project DICOM.NetCore: it is built in net standard and therefore supports as much platforms as it will be possible in future. It perfectly integrates efferent codecs. Removing the shproj here with refactoring to a minimum would only be copying all the files from the shared project into this DICOm.NetCore project. 
The only different/new thing would be a test suite to verify rendering, but thats what I would also like to add to the current 4.0.x version.
Did I miss something?

the next part, namely refactoring etc, is something I already started in netstandard-branch. Did you take a closer look at the code so far?

**Comment by amoerie** [comment]
Hmm fair enough, I suppose the Dicom.NetCore project will suffice for another while. In that case I agree with you: refactoring ahoy!

I've quickly perused your branch already, but it would help if you have a summary somewhere with the changes, because right now they all count as "new files" and it's impossible for me to easily see what you've modified.

As far as I'm concerned, you don't need to merge to the development branch yet, I can open pull requests to your netstandard branch. Before I do so though, do you think it's in a stable enough state already or are you still moving big parts around? I would like to avoid huge merge conflicts if possible.

**Comment by amoerie** [comment]
This issue is obsolete, we now have a full clone of the solution that uses .NET Standard under the hood.
